### PR TITLE
[CSL-1442/CSL-1443] Improve `computeSharesDistr` and tests. Improve `Arbitrary ValidRichmenStakes`. 

### PR DESCRIPTION
**Pull request by pva701**
* Arbitrary instance for `ValidRichmenStake` is rewritten.
* `isFairDistr` uses knapsack algorithm instead of brute force to provide opportunity to run big tests (n > 20)
* `computeSharesDistr` uses knapsack algorithm to provide strict guarantees on a difference between real and generated distribution.

I would like to implement in this issue (or maybe in next issues):
* add optimization in knapsack part: if we achieve inaccuracy `sharesDistrInaccuracy` on some step of algorithm then break and return it
* I thought of the worst case of the test for `computeSharesDistr` and got some results, I would like to add some "clever" tests about it.

### CODE REVIEW

**Comment by pva701** [comment]
@gromakovsky welcome to diss

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 1]
```diff
@@ -136,6 +139,58 @@ computeParticipants (HS.toMap -> richmen) = flip HM.intersection richmen
...
+        forM_ weights $ \w -> when (w + distr <= totalDistr) $ do
+            relax dpMax coins w (w + distr) (>)
+            relax dpMin coins w (w + distr) (<)
+
+    let computeError :: Word16 -> Int -> Double
+        computeError i sCoins = abs $
+            fromIntegral i / fromIntegral totalDistr -
+            fromIntegral sCoins / fromIntegral totalCoins
+    let selectMax = flip execStateT 0.0 $ forM_ [0..totalDistr] $ \i -> do
+            if | i < halfDistr -> do
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: I suppose this `if` is redundant, you can use guard without this `if`.

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 2]
```diff
@@ -32,13 +33,15 @@ module Pos.Ssc.GodTossing.Toss.Base
...
-import           Control.Monad.State             (get, put)
 import           Data.Containers                 (ContainerKey, SetContainer (notMember))
 import qualified Data.HashMap.Strict             as HM
 import qualified Data.HashSet                    as HS
 import qualified Data.List.NonEmpty              as NE
 import           Formatting                      (ords, sformat, (%))
 import           System.Wlog                     (logWarning)
+import           Data.Array.MArray               (newArray, readArray, writeArray)
+import           Data.Array.ST                   (STUArray)
+import           Control.Monad.ST                (ST, runST)
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: Did you disable stylish-haskell?

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 3]
```diff
@@ -144,78 +199,54 @@ computeSharesDistrPure
...
-        let initS = sum init
-        let initDelta = calcSumError (map (`divRat` initS) init) portions
-        (_, _, res) <-
-                execStateT (compute fromX toX epsilon portions) (initDelta, initS, init)
-        pure $ HM.fromList $ zip keys res
+        case compute fromX toX 0 of
+            Nothing -> throwError $ TossInternallError "There is no appropriate distribution"
+            Just res -> pure $ HM.fromList $ zip keys res
   where
+    sharesDistrInaccuracyR = sharesDistrInaccuracy :: Double
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: Why `R`?

Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: `R` means `Real gansta nigga`

Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: I meant `Real ~ Double`

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 2 above)
Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: it works as fuck

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 3 above)
Comment (Nafanya) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: R for RIP

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 4]
```diff
@@ -144,78 +199,54 @@ computeSharesDistrPure
...
     | null richmen = pure mempty
     | otherwise = do
-        let total :: Word64
-            total = sum $ map unsafeGetCoin $ toList richmen
-        when (total == 0) $
+        when (totalCoins == 0) $
             throwError $ TossInternallError "Richmen total stake equals zero"
-        let epsilon = sharesDistrInaccuracy::Rational
         -- We accept error in computation = 0.05,
         -- so stakeholders must have at least 55% of stake (for reveal secret) in the worst case
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: Why is this comment here, btw?

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 5]
```diff
@@ -144,78 +199,54 @@ computeSharesDistrPure
...
             throwError $ TossInternallError "Richmen stakes less than threshsold"
 
-        let init = normalize $ multPortions portions toX
-        let initS = sum init
-        let initDelta = calcSumError (map (`divRat` initS) init) portions
-        (_, _, res) <-
-                execStateT (compute fromX toX epsilon portions) (initDelta, initS, init)
-        pure $ HM.fromList $ zip keys res
+        case compute fromX toX 0 of
+            Nothing -> throwError $ TossInternallError "There is no appropriate distribution"
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: I don't find it acceptable. If we can't find a distribution, we must use some default (e. g. just give 1 share for everybody).

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 6]
```diff
@@ -144,78 +199,54 @@ computeSharesDistrPure
...
-        when (total == 0) $
+        when (totalCoins == 0) $
             throwError $ TossInternallError "Richmen total stake equals zero"
-        let epsilon = sharesDistrInaccuracy::Rational
         -- We accept error in computation = 0.05,
         -- so stakeholders must have at least 55% of stake (for reveal secret) in the worst case
         let mpcThreshold = toRational (getCoinPortion threshold) / toRational coinPortionDenominator
-        let fromX = 1
+        let fromX = ceiling $ 1.0 / minimum portions
         let toX = truncate $ toRational (3::Int) / mpcThreshold
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: I think `toRational (3::Int)` is redundant here (can be just `3`).

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 7]
```diff
@@ -144,78 +199,54 @@ computeSharesDistrPure
...
-                execStateT (compute fromX toX epsilon portions) (initDelta, initS, init)
-        pure $ HM.fromList $ zip keys res
+        case compute fromX toX 0 of
+            Nothing -> throwError $ TossInternallError "There is no appropriate distribution"
+            Just res -> pure $ HM.fromList $ zip keys res
   where
+    sharesDistrInaccuracyR = sharesDistrInaccuracy :: Double
+
+    coins :: [CoinUnsafe]
+    coins = map (fromIntegral . unsafeGetCoin) (toList richmen)
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: Hm, previously you didn't want to rely on the fact that `HM.toList`, `toList`, `HM.elems` and `HM.keys` all return the same order.
Now you rely on this fact here. So probably it makes sense to use `HM.keys` above.

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 8]
```diff
@@ -136,6 +139,58 @@ computeParticipants (HS.toMap -> richmen) = flip HM.intersection richmen
...
+
+-- ATTENTION: IMPERATIVE CODE! PROTECT YOUR EYES! --
+-- | Compute inaccuracy betwee real distribution and generared.
+-- It take O(totalDistr * length consNDistr) time.
+computeDistrInaccuracy :: (CoinUnsafe, Word16) -> [(CoinUnsafe, Word16)] -> Double
+computeDistrInaccuracy (totalCoins, totalDistr) coinsNDistr = runST $ do
+    let halfDistr = totalDistr `div` 2 + 1
+    let invalid = totalCoins + 1 :: Int
+    dpMax <- newArray (0, totalDistr) (-invalid) :: ST s (Knapsack s)
+    dpMin <- newArray (0, totalDistr) invalid    :: ST s (Knapsack s)
```
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: Can you comment what exactly is stored there?

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 3 above)
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: Actually it is redundant, because the type will be inferred later.

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 8 above)
Comment (gromakovsky) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: And add more comments to this function please.

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 3 above)
Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: I know, but I didn't want inferring every time

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 9]
```diff
@@ -136,6 +139,58 @@ computeParticipants (HS.toMap -> richmen) = flip HM.intersection richmen
 sharesDistrInaccuracy :: Fractional a => a
 sharesDistrInaccuracy = 0.05
 
+-- | Internal type for work with Coin.
+type CoinUnsafe = Int
```
Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: don't give a fuck, `Int` fits into 2^64 on my machine :peka:
but ok, I will think of it

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 1 above)
Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: it doesn't work(

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs (hunk 7 above)
Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: good point, I forgot about it

**Comment by pva701** [comment]
@gromakovsky I couldn't decide commit `[CSL-1442] Add knapsack optimization` really whether need

it reduces time of tests from ~44sec to 35sec, but the code of `isDistrInaccuracyAcceptable` is more sophisticated

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs [hunk 10]
```diff
@@ -87,81 +89,36 @@ testMpcThdPortition = unsafeCoinPortionFromDouble testMpcThd
...
     isLimitedBy mx sd &&
-    -- (2) Distribution is fair (see explanation of @isDistrFair@)
+    -- (4) Distribution is fair (see explanation of @isDistrFair@)
     isDistrFair rs sd
+  where
+    totalCoins = sumCoins $ HM.elems rs
+    minStake = mkCoin . ceiling $ (fromIntegral totalCoins) * testMpcThd
 
 isDistrReasonableMax :: RichmenStakes -> Either TossVerFailure SharesDistribution -> Bool
 isDistrReasonableMax = isDistrReasonable $ truncate $ toRational (3::Int) / toRational testMpcThd
```
Comment (gromakovsky) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: It would be good not to duplicate this `3 / toRational thd` magic.

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs [hunk 11]
```diff
@@ -87,81 +89,36 @@ testMpcThdPortition = unsafeCoinPortionFromDouble testMpcThd
...
-        let !(er, firstCase, secondCase) = fairBrute distrs 0 0
-        er < sharesDistrInaccuracy &&
-            toRational firstCase < totalCases * fivePerc &&
-            toRational secondCase < totalCases * fivePerc
+isDistrFair (HM.map unsafeGetCoin -> rs) sd = do
+    let !totalDistr = sum sd
+    let !totalCoins = sum rs
+    let stakeholders = HM.keys rs
+    let coinsNDistr = map (first fromIntegral . findStk) stakeholders
+    isDistrInaccuracyAcceptable (fromIntegral totalCoins, totalDistr) coinsNDistr
```
Comment (gromakovsky) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: The problem now is that we rely on `isDistrInaccuracyAcceptable` correctness, but we don't know whether it's correct.
It would be good to add some tests for it. Probably very simple, just hardcode some numbers (close to corner cases).

File: godtossing/Pos/Ssc/GodTossing/Toss/Base.hs [hunk 12]
```diff
@@ -144,78 +227,50 @@ computeSharesDistrPure
...
-    divRat :: (Real a, Real b) => a -> b -> Rational
-    divRat x y = toRational x / toRational y
+    compute :: Word16 -> Word16 -> Word16 -> Maybe [Word16]
+    compute !x !toX !prevSum
+        | x > toX = Nothing
+        | otherwise = do
+            let curDistrN = multPortions x
+            let curDistr = normalize curDistrN
+            let !s = sum curDistr
+            if s == prevSum then compute (x + 1) toX prevSum
```
Comment (pva701) [inline godtossing/Pos/Ssc/GodTossing/Toss/Base.hs]: 1. it is. when you add `1` to x you add to each shares distribution `i` real `p_i`, and if each number isn't exceeded next integer number then each number equals itself previous value - a sum of distributions is same 
2. no, it's not possible. because >each number equals itself previous value

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs (hunk 11 above)
Comment (pva701) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: yep, also, problem is that we rely on `isDistrInaccuracyAcceptable` in the code and in the tests.
it's dangerous function....

**Comment by pva701** [comment]
I optimized tests of `computeSharesDistr` as fuck. Now it works `12s` instead `44s` on my machine

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs [hunk 13]
```diff
@@ -72,7 +77,22 @@ spec = describe "computeSharesDistr" $ do
     \ to test validity of @computeSharesDistr@"
     twentyRichmen1Desc = "20 richmen with similar stake to test fairness of generated distribution"
     twentyRichmen2Desc = "20 richmen with similar stake to test fairness of generated distribution"
-    validateFairnessDesc = "Given a valid richmen, validate fairness"
+    validateReasonableDesc = "Given a valid richmen, validate fairness and some reasonable statements"
+
+    distrEqDesc = "Distribution fits coins"
```
Comment (gromakovsky) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: I have no idea what this sentence means.

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs [hunk 14]
```diff
@@ -72,7 +77,22 @@ spec = describe "computeSharesDistr" $ do
...
+    distrErrorType1Desc = "Distribution tells we can't reveal a secret, but coins tell we can"
+    distrErrorType2Desc = "Distribution tells we can reveal a secret, but coins tell we can't"
+    distrHalfSum1Desc =
+        "Distribution equals @sumDistr / 2@ , sum of coins is greater > @sumCoins / 2@.\
```
Comment (gromakovsky) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: `greater >`?

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs [hunk 15]
```diff
@@ -292,3 +243,22 @@ lrcConsistency =
...
+distrFits = isDistrInaccuracyAcceptable [(10, 1), (20, 2), (30, 3)]
+
+distrErrorType1 :: Bool
+distrErrorType1 = not $ isDistrInaccuracyAcceptable [(20, 2), (40, 2), (10, 1), (20, 2), (20, 2)]
+
+distrErrorType2 :: Bool
+distrErrorType2 = not $ isDistrInaccuracyAcceptable [(10, 2), (30, 4), (20, 2), (30, 3)]
+
+distrHalfSum1 :: Bool
+distrHalfSum1 = isDistrInaccuracyAcceptable [(20, 2), (41, 3), (10, 1), (20, 2), (20, 2)]
```
Comment (gromakovsky) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: It would be good to write somewhere that `61 / 111 = 0.5495495495495496`.

File: test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs (hunk 13 above)
Comment (pva701) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: slabak

Comment (pva701) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: maybe "Распределение аппроксимирует монеты"

Comment (gromakovsky) [inline test/Test/Pos/Ssc/GodTossing/ComputeSharesSpec.hs]: Dunno, I approved anyway.
